src/lib/x509/x509_ext.cpp

```diff
@@ -868,6 +877,582 @@ const std::string& TNAuthList::Entry::telephone_number() const {
+
+   if(addr_family_length != 2 && addr_family_length != 3) {
+      throw Decoding_Error("(S)AFI can only contain 2 or 3 bytes");
+   }
+
+   m_afi = (addr_family[0] << 8) | addr_family[1];
+
+   if(addr_family_length == 3) {
+      m_safi = addr_family[2];
+   }
+
+   if(m_afi == 1) {
+      IPAddressChoice<Version::IPv4> addr_choice;
+      obj_ber.decode(addr_choice);
+      m_ip_addr_choice = addr_choice;
+   } else if(m_afi == 2) {
+      IPAddressChoice<Version::IPv6> addr_choice;
+      obj_ber.decode(addr_choice);
+      m_ip_addr_choice = addr_choice;
+   } else {
+      throw Decoding_Error("Only AFI IPv4 and IPv6 are supported.");
+   }
+}
+
+template <IPAddressBlocks::Version V>
+void IPAddressBlocks::IPAddressChoice<V>::encode_into(Botan::DER_Encoder& into) const {
+   if(m_ip_addr_ranges.has_value()) {
+      into.start_sequence().encode_list(m_ip_addr_ranges.value()).end_cons();
+   } else {
+      into.encode_null();
+   }
+}
+
+template <IPAddressBlocks::Version V>
+void IPAddressBlocks::IPAddressChoice<V>::decode_from(Botan::BER_Decoder& from) {
+   ASN1_Type next_tag = from.peek_next_object().type_tag();
+
+   if(next_tag == ASN1_Type::Null) {
+      from.decode_null();
+      m_ip_addr_ranges = std::nullopt;
+   } else if(next_tag == ASN1_Type::Sequence) {
+      std::vector<IPAddressOrRange<V>> ip_ranges;
+      from.decode_list(ip_ranges);
+      m_ip_addr_ranges = ip_ranges;
+   } else {
+      throw Decoding_Error(fmt("Unexpected type for IPAddressChoice {}", static_cast<uint32_t>(next_tag)));
+   }
+}
+
+template <IPAddressBlocks::Version V>
+void IPAddressBlocks::IPAddressChoice<V>::sort_and_merge() {
+   if(!m_ip_addr_ranges.has_value()) {
+      return;
+   }
+   std::vector<IPAddressOrRange<V>> current = m_ip_addr_ranges.value();
+   if(current.empty()) {
+      return;
+   }
+   bool changed = true;
+   while(changed) {
+      changed = false;
+      // sort by the minimum address
+      std::sort(current.begin(), current.end(), [](IPAddressOrRange<V>& a, IPAddressOrRange<V>& b) {
+         return a.min() < b.min();
+      });
+      for(size_t i = 0; i < current.size() - 1; i++) {
+         IPAddressOrRange<V> a = current[i];
+         IPAddressOrRange<V> b = current[i + 1];
+         // only ++ is implemented for IPAddress<V>, instead of +
+         if(b.min() <= a.max() || b.min() == (++a.max())) {
+            // erase old a and b
+            current.erase(current.begin() + i, current.begin() + i + 2);
+            current.insert(current.begin() + i, IPAddressOrRange<V>(a.min(), std::max(a.max(), b.max())));
+            // look again at the new element we just created, might need to be merged with others still
+            i -= 1;
+            changed = true;
+         }
+      }
+   }
+   m_ip_addr_ranges = current;
+}
+
+template <IPAddressBlocks::Version V>
+void IPAddressBlocks::IPAddressOrRange<V>::encode_into(Botan::DER_Encoder& into) const {
+   const uint8_t version_octets = static_cast<uint8_t>(V);
```

Comment:
In general use `size_t` for lengths unless there is a specific reason to use `uintX_t`

Also using `size_t` here avoids having to cast it later in the function